src/DataStructures/Tensor/Expressions/Contract.hpp

```diff
@@ -168,29 +189,85 @@ struct TensorContract
+ * encodings, this function looks to see if it can find a pair of values that
+ * encode one generic index and the generic index with opposite valence, such as
+ * `ti_A_t` and `ti_a_t`. This denotes a pair of indices that will need to be
+ * contracted. If there exists more than one such pair of indices in the
+ * expression, the first pair of values found will be returned.
+ *
+ * For example, if we have tensor \f${R^{ab}}_{ab}\f$ represented by the tensor
+ * expression, `R(ti_A, ti_B, ti_a, ti_b)`, then this will return the positions
+ * of the pair of values encoding `ti_A_t` and `ti_a_t`, which would be (0, 2)
+ *
+ * @param tensorindex_values the TensorIndex values of a tensor expression
+ * @return the positions of the first pair of TensorIndex values to contract
  */
-template <size_t ReplacedArg1, size_t ReplacedArg2, typename T, typename X,
-          typename Symm, typename IndexList, typename Args>
-SPECTRE_ALWAYS_INLINE auto contract(
-    const TensorExpression<T, X, Symm, IndexList, Args>& t) {
-  return TensorContract<tmpl::size_t<ReplacedArg1>,
-                        tmpl::size_t<ReplacedArg2>, T, X, Symm, IndexList,
-                        Args>(~t);
+template <size_t NumIndices>
+SPECTRE_ALWAYS_INLINE static constexpr std::pair<size_t, size_t>
+get_first_index_positions_to_contract(
+    const std::array<size_t, NumIndices>& tensorindex_values) noexcept {
+  for (size_t i = 0; i < tensorindex_values.size(); ++i) {
+    const size_t current_value = gsl::at(tensorindex_values, i);
+    const size_t opposite_value_to_find =
+        get_tensorindex_value_with_opposite_valence(current_value);
+    for (size_t j = i + 1; j < tensorindex_values.size(); ++j) {
+      if (opposite_value_to_find == gsl::at(tensorindex_values, j)) {
+        // We found both the lower and upper version of a generic index in the
+        // list of generic indices, so we return this pair's positions
+        return std::pair{i, j};
+      }
+    }
+  }
+  // We couldn't find a single pair of indices that needs to be contracted
+  return std::pair{std::numeric_limits<size_t>::max(),
+                   std::numeric_limits<size_t>::max()};
 }
 
-namespace detail {
-// Helper struct to allow contractions by using repeated indices in operator()
-// calls to tensor.
-template <template <typename> class TE, typename ReplacedArgList, size_t I,
-          typename TotalContracted, typename T>
-SPECTRE_ALWAYS_INLINE static constexpr auto fully_contract(const T& t) {
-  if constexpr (I == 2 * (TotalContracted::value - 1)) {
-    return contract<ti_contracted_t<I>::value, ti_contracted_t<I + 1>::value>(
-        TE<ReplacedArgList>(t));
+/*!
+ * \ingroup TensorExpressionsGroup
+ * \brief Creates a contraction expression from a tensor expression if there are
+ * any indices to contract
+ *
+ * \details If there are no indices to contract, the input TensorExpression is
+ * simply returned. If not, a contraction expression is created for contracting
```

Comment:
[nitpick] "If not" -> "Otherwise" (because the prior sentence describes the negative case, it's confusing to start this one with "if not")

Reply:
Done!